tests/check_framework/test_translation.py

```diff
@@ -33,9 +34,66 @@ def test_invalid_language_code(self):
             # 'sr@latin',      # locale instead of language tag.
         )
-        for tag in tags:
+
+    def test_valid_language_code(self):
+        for tag in self.valid_tags:
+            with self.subTest(tag), override_settings(LANGUAGE_CODE=tag):
+                self.assertEqual(check_setting_language_code(None), [])
+
+    def test_invalid_language_code(self):
+        msg = 'You have provided an invalid value for the LANGUAGE_CODE setting: %s.'
+        for tag in self.invalid_tags:
             with self.subTest(tag), override_settings(LANGUAGE_CODE=tag):
                 result = check_setting_language_code(None)
-                self.assertEqual(result, [E001])
+                error = copy(E001)
+                error.msg = error.msg.format(tag)
+                self.assertEqual(result, [error])
                 self.assertEqual(result[0].id, 'translation.E001')
-                self.assertEqual(result[0].msg, 'You have provided an invalid value for the LANGUAGE_CODE setting.')
+                self.assertEqual(result[0].msg, msg % tag)
+
+    def test_valid_languages(self):
+        for tag in self.valid_tags:
+            with self.subTest(tag), override_settings(LANGUAGES=[(tag, tag)]):
+                self.assertEqual(check_setting_languages(None), [])
+
+    def test_invalid_languages(self):
+        msg = 'You have provided an invalid language code in the LANGUAGES setting: %s.'
+        for tag in self.invalid_tags:
+            with self.subTest(tag), override_settings(LANGUAGES=[(tag, tag)]):
+                result = check_setting_languages(None)
+                error = copy(E002)
+                error.msg = error.msg.format(tag)
+                self.assertEqual(result, [error])
+                self.assertEqual(result[0].id, 'translation.E002')
+                self.assertEqual(result[0].msg, msg % tag)
+
+    def test_valid_languages_bidi(self):
+        for tag in self.valid_tags:
+            with self.subTest(tag), override_settings(LANGUAGES_BIDI=[tag]):
+                self.assertEqual(check_setting_languages_bidi(None), [])
+
+    def test_invalid_languages_bidi(self):
+        msg = 'You have provided an invalid language code in the LANGUAGES_BIDI setting: %s.'
+        for tag in self.invalid_tags:
+            with self.subTest(tag), override_settings(LANGUAGES_BIDI=[tag]):
+                result = check_setting_languages_bidi(None)
+                error = copy(E003)
+                error.msg = error.msg.format(tag)
+                self.assertEqual(result, [error])
+                self.assertEqual(result[0].id, 'translation.E003')
+                self.assertEqual(result[0].msg, msg % tag)
+
+    def test_inconsistent_language_settings(self):
+        msg = 'You have provided a value for the LANGUAGE_CODE setting that is not in the LANGUAGES setting.'
+        with override_settings(LANGUAGE_CODE='fr', LANGUAGES=[('en', 'English')], LANGUAGES_BIDI=[]):
```

Comment:
Does it require an overriding of  `LANGUAGES_BIDI`?

Reply:
Yes, otherwise `translation.W005` is also triggered.

Reply:
Ah, ok so maybe we can check both in one call, i.e.
```python
self.assertEqual(check_language_settings_consistent(None), [
    Error(
        ...
    ),
    Warning(
        ...
    ),
])
```

Reply:
I feel this is clearer kept separate.